File: python/scripts/week-02/demo-01-prose-vs-structured.py
"""
Demo 1: Unstructured request → crash, then Schema → guaranteed.

The moderator's "Demo 1": ask the LLM for the fields you want — but with NO
schema. The model gives you those fields back as prose, so json.loads() crashes.
Then add a schema constraint (analyze_pr) and get a typed object that ALWAYS
parses. A request is not a contract.

Run: python scripts/week-02/demo-01-prose-vs-structured.py
"""
import json
import os
import sys
import warnings

warnings.filterwarnings("ignore", message=".*Pydantic serializer.*")
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from langchain_core.messages import HumanMessage
from src.llm import get_llm
from src.llm_functions import analyze_pr
from src.config import DEVBUDDY_MODEL

PR_TITLE = "Fix login redirect loop in auth-service"
PR_DIFF = (
    "Changed session token validation in src/auth.py lines 42-58.\n"
    "Previously expired tokens caused infinite redirect for 15% of users.\n"
    "Now returns 401 with clear error. No DB changes. Rollback: revert commit.\n"
    "Files: src/auth.py, tests/test_auth.py"
)

print("=" * 70)
print("  Demo 1: Prose → crash, then Schema → success")
print("=" * 70)
print(f"  Model: {DEVBUDDY_MODEL}")
print()
print(f"  INPUT: {PR_TITLE}")
for line in PR_DIFF.split("\n"):
    print(f"         {line}")
print()

# ═══════════════════════════════════════════════════════════════
# Step 1: plain prompt → prose
# ═══════════════════════════════════════════════════════════════
print("-" * 70)
print("  STEP 1: ask for the fields (no schema) → prose")
print("-" * 70)
print()

llm = get_llm(temperature=0.0)
raw = llm.invoke([HumanMessage(content=(
    f"Analyze this PR and return its severity, a one-sentence summary, "
    f"and the affected files.\n\n"
    f"PR Title: {PR_TITLE}\n\nDiff:\n{PR_DIFF}"
))]).content

print("  Raw response:")
for line in raw.strip().split("\n"):
    print(f"  | {line}")
print()

# ═══════════════════════════════════════════════════════════════
# Step 2: json.loads → crash
# ═══════════════════════════════════════════════════════════════
print("-" * 70)
print("  STEP 2: json.loads(raw) → crash")
print("-" * 70)
print()

try:
    data = json.loads(raw)
    print(f"  ✅ parsed unexpectedly: {list(data.keys())}")
except json.JSONDecodeError as e:
    print(f"  ❌ CRASHED: {e}")
    print()
    print("  You asked for the severity, summary, and affected files.")
    print("  The model gave them to you — as prose. With no schema,")
    print("  the model picks the format, and it picked prose.")
    print("  Free text breaks pipelines. Let's fix it.")
print()

# ═══════════════════════════════════════════════════════════════
# Step 3: schema constraint → typed object
# ═══════════════════════════════════════════════════════════════
print("-" * 70)
print("  STEP 3: analyze_pr (schema-constrained) → typed object")
print("-" * 70)
print()

result = analyze_pr(PR_TITLE, PR_DIFF, temperature=0.0)
print(f"  type: {type(result).__name__}  ← typed object, not prose!")
print()
print(result.model_dump_json(indent=2))
print()

print("=" * 70)
print("  The LLM is now a typed function.")
print("  Input → BuildCheck. Your code consumes it directly.")
print("=" * 70)

File: nodejs/scripts/week-02/demo-01-prose-vs-structured.js
/**
 * Demo 1: Prose → crash, then Schema → success.
 *
 * The moderator's "Demo 1": call the LLM with a plain prompt and get prose
 * back, feed that prose to JSON.parse() → crash. Then add a schema constraint
 * (analyzePr) and get a typed object back.
 *
 * Run: node scripts/week-02/demo-01-prose-vs-structured.js
 */
import { HumanMessage } from "@langchain/core/messages";
import { getLlm } from "../../src/llm.js";
import { analyzePr } from "../../src/llm_functions.js";

const PR_TITLE = "Fix login redirect loop in auth-service";
const PR_DIFF = [
  "Changed session token validation in src/auth.py lines 42-58.",
  "Previously expired tokens caused infinite redirect for 15% of users.",
  "Now returns 401 with clear error. No DB changes. Rollback: revert commit.",
  "Files: src/auth.py, tests/test_auth.py",
].join("\n");

console.log("=".repeat(70));
console.log("  Demo 1: Prose → crash, then Schema → success");
console.log("=".repeat(70));
console.log();
console.log(`  INPUT: ${PR_TITLE}`);
for (const line of PR_DIFF.split("\n")) {
  console.log(`         ${line}`);
}
console.log();

// ═══════════════════════════════════════════════════════════════
// Step 1: plain prompt → prose
// ═══════════════════════════════════════════════════════════════
console.log("-".repeat(70));
console.log("  STEP 1: plain prompt → prose");
console.log("-".repeat(70));
console.log();

const llm = getLlm({ temperature: 0.0 });
const raw = (
  await llm.invoke([
    new HumanMessage(`Summarize this PR: ${PR_TITLE}\n\n${PR_DIFF}`),
  ])
).content;

console.log("  Raw response:");
for (const line of raw.trim().split("\n")) {
  console.log(`  | ${line}`);
}
console.log();

// ═══════════════════════════════════════════════════════════════
// Step 2: JSON.parse → crash
// ═══════════════════════════════════════════════════════════════
console.log("-".repeat(70));
console.log("  STEP 2: JSON.parse(raw) → crash");
console.log("-".repeat(70));
console.log();

try {
  const data = JSON.parse(raw);
  console.log(`  ✅ parsed unexpectedly: ${Object.keys(data).join(", ")}`);
} catch (e) {
  console.log(`  ❌ CRASHED: ${e.message}`);
  console.log();
  console.log("  This is code slop. Free text breaks pipelines. Let's fix it.");
}
console.log();

// ═══════════════════════════════════════════════════════════════
// Step 3: schema constraint → typed object
// ═══════════════════════════════════════════════════════════════
console.log("-".repeat(70));
console.log("  STEP 3: analyzePr (schema-constrained) → typed object");
console.log("-".repeat(70));
console.log();

const result = await analyzePr({
  title: PR_TITLE,
  diff: PR_DIFF,
  temperature: 0.0,
});

console.log(`  type: ${result.constructor.name}  ← typed object, not prose!`);
console.log();
console.log(JSON.stringify(result, null, 2));
console.log();

console.log("=".repeat(70));
console.log("  The LLM is now a typed function.");
console.log("  Input → BuildCheck. Your code consumes it directly.");
console.log("=".repeat(70));
